### ISSUE TITLE
Implement Object Responses for Corners and Sides

### ISSUE THREAD

**Issue by terracoda** [phetsims/quadrilateral#130]
The phrase patterns for the object responses for corners and sides are laid out in 2 tables in the design document.

**Object Response Phrase Pattern Rules (Parallelograms):**
- First provide details about the opposite corner or side, then call out important details:
    - When corner is a right angle, prepend “right angle”
    - When all corners/sides are equal (squares, rhombuses, rectangles), if POSSIBLE use a shorter phrase:
        - [[equal to opposite and adjacent corners]]
        - [[equal to opposite and adjacent sides]]
- Second describe the corner/side in relation to its adjacents
    - All opposites are equal for Parallelograms, so I would like to add [[equal]] into the adjacent phrase, if that is not too much trouble
        - wider than adjacent [[equal]] corners
        - smaller than adjacent [[equal]] corners 

**Object Response Phrase Pattern Rules (non-parallelograms):**
- First provide details about the opposite corner or side, then call out important details:
    - When corner is a right angle, prepend “right angle”
- Second describe the corner/side in relation to its adjacents
    - When adjacent corners/sides are different than compared corner/side but similar in comparison use a simplified binary comparative scale (smaller/wider or shorter/longer) rather than the full comparative scale
        - “smaller/wider than adjacent corners”
        - “shorter/longer than adjacent sides”
    - When one adjacent corner or side is equal, call it out and do not describe the other adjacent corner/side
        - [[equal to one adjacent corner]]
        - [[equal to one adjacent side]]
    - When adjacent corners/sides are all together different from the compared corner/side and from each other describe the corner/side as “not equal to” adjacent corners/sides
        - [[not equal to adjacent corners]]
        - [[not equal to adjacent sides]]

**Comment by terracoda** [comment]
Here are the links to the examples in the Voicing Design Doc
- [Table: Possible Object Responses for Corners & Sides (Parallelograms)](https://docs.google.com/document/d/1jXayebAWnnNzsT3l6o72YPw4-YtiQaHQNuAi64eiguc/edit#heading=h.2apgyax1g10t)
- [Table: Possible Object Responses for Corners & Sides (non-Parallelograms)](https://docs.google.com/document/d/1jXayebAWnnNzsT3l6o72YPw4-YtiQaHQNuAi64eiguc/edit#heading=h.ic6peq8txg0e)

@jessegreenberg, regarding the scales for the parameters...
1. To describe the opposite corner/side use the same qualitative comparative scales as we do for dynamic state descriptions:
    - [Comparative Corner Size Parameters](https://docs.google.com/document/d/1jXayebAWnnNzsT3l6o72YPw4-YtiQaHQNuAi64eiguc/edit#heading=h.fzsgp4ejoxgp)
    - [Comparative Side Length Parameters](https://docs.google.com/document/d/1jXayebAWnnNzsT3l6o72YPw4-YtiQaHQNuAi64eiguc/edit#heading=h.surzggpll38r)
2. To describe the adjacent corners/sides, there are 3 cases:
    - When they can be described together, use a simplified binary scale 
    - When one is equal, just describe the equal corner or side
    - When both are different, describe the corner or side as "not equal" to them

**Comment by terracoda** [comment]
@jessegreenberg, assigning over to you. Let me know if you have any questions.

**Comment by terracoda** [comment]
I think these phrase patterns will work for Interactive Description, but I will open a separate issue for that.

**Comment by terracoda** [comment]
@jessegreenberg, one thing I didn't do in the design doc was put curly braces around parameters. I'll do that now.

**Comment by terracoda** [comment]
Ok, @jessegreenberg, I added the curly braces and actually refined the rules slightly (sorry!)

[Table: Possible Object Responses for Corners & Sides (Parallelograms)](https://docs.google.com/document/d/1jXayebAWnnNzsT3l6o72YPw4-YtiQaHQNuAi64eiguc/edit#heading=h.2apgyax1g10t)
**Goal of Design:** Tabbing through all corners and sides ideally provides enough details to determine what the current shape is.

**Object Response Phrase Pattern Rules (Parallelograms):**
1. First provide details about the opposite corner or side, then call out important details:
    - When corner is a right angle, prepend “right angle”
    - When all corners/sides are equal (squares, rhombuses, rectangles), if POSSIBLE use a shorter combined phrase instead of the two separate phrases:
        - [[equal to opposite and adjacent corners]]
        - [[equal to opposite and adjacent sides]]
2. Second describe the corner/side in relation to its adjacents, and since all opposites are equal for Parallelograms, I would like to do two things different than what we do for the other shapes IF POSSIBLE:
    - use the full comparative scales rather than the simplified comparison scales
        - {{somewhat wider than}} adjacent [[equal]] corners
        - {{much shorter than}} adjacent [[equal]] sides 
    - add [[equal]] into the adjacent phrase 
        - adjacent {{[[equal]]}} corners
        - adjacent {{[[equal]]}} sides 

[Table: Possible Object Responses for Corners & Sides (non-Parallelograms)](https://docs.google.com/document/d/1jXayebAWnnNzsT3l6o72YPw4-YtiQaHQNuAi64eiguc/edit#heading=h.ic6peq8txg0e)
**Goal of Design:** 
- Tabbing through all corners and sides ideally provides enough details to determine what the current shape is, or at least know that it is not a parallelogram. 
- With some deduction, a learner can figure out if it is better to play with a corner or play with a side to find a parallelogram.

**Object Response Phrase Pattern Rules (non-parallelograms):**
1. First provide details about the opposite corner or side, then call out important details:
    - When corner is a right angle, prepend “right angle”
2. Second describe the corner/side in relation to its adjacents, but use a simplified comparative scale rather than the full comparative scale
    - When adjacent corners/sides are different than compared corner/side but similar in comparison to each other
        - “smaller/wider than adjacent corners”
        - “shorter/longer than adjacent sides”
    - When one adjacent corner or side is equal, call it out and do not describe the other adjacent corner/side
        - [[{{equal to one}} adjacent corner]]
        - [[equal to one}} adjacent side]]
    - When adjacent corners/sides are all together different from the compared corner/side and from each other describe the corner/side as “not equal to” adjacent corners/sides
        - [[{{not equal to}} adjacent corners]]
        - [[{{not equal to}} adjacent sides]]

**Comment by terracoda** [comment]
I also spell out the Simplified Comparative Scales that can be used for describing adjacent corners & sides in non-parallelograms

[Table: Simplified Comparative Corner Size Parameters](https://docs.google.com/document/d/1jXayebAWnnNzsT3l6o72YPw4-YtiQaHQNuAi64eiguc/edit#heading=h.t9g7ilxdyv2)
![image](https://user-images.githubusercontent.com/1902116/164235462-d0b40c5e-f8ac-4577-a530-5676ec22e761.png)


[Table: Simplified Comparative Side Length Parameters](https://docs.google.com/document/d/1jXayebAWnnNzsT3l6o72YPw4-YtiQaHQNuAi64eiguc/edit#heading=h.e8pilbnrr3pb)
![image](https://user-images.githubusercontent.com/1902116/164235357-b7f3f1f2-385a-46df-b7f6-9134d7408fcd.png)

**Comment by terracoda** [comment]
@jessegreenberg, I reviewed the rules and the examples in the design document. The examples should be clean.

I did adjust the rules slightly....

**Object Response Phrase Pattern Rules (Parallelograms):**

- First provide details about the opposite corner or side, then call out important details:
    - When corner is a right angle, prepend “right angle”
- Second describe the corner/side in relation to its adjacents, and since all opposites are equal for Parallelograms, I would like to:
    - use the [full comparative scales](https://docs.google.com/document/d/1jXayebAWnnNzsT3l6o72YPw4-YtiQaHQNuAi64eiguc/edit#heading=h.55vefl1yzfor) & 
    - add have an “equal” parameter for the adjacent corners/sides
        - {{somewhat wider than}} adjacent {{equal}} corners
        - {{much shorter than}} adjacent {{equal}} sides 
- When “equal to opposite” corner/side and “equal to adjacent” corner/side - CAN we leave off the third “equal”
- Note: See the[ simplified comparative scale](https://docs.google.com/document/d/1jXayebAWnnNzsT3l6o72YPw4-YtiQaHQNuAi64eiguc/edit#heading=h.t9g7ilxdyv2) for describing adjacent corners and adjacent sides is not relevant to Parallelograms.

**Object Response Phrase Pattern Rules (non-parallelograms):**

- First provide details about the opposite corner or side using the full comparative scale, and call out important details:
    - When corner is a right angle, prepend “right angle” (same as in parallelograms)
    _- I think we might need to call out when sides start in parallel._
        - " {{parallel to and}} {{somewhat longer than}} opposite side"
- Second describe the corner/side in relation to its adjacents, but when the adjacent corners/sides are different from each other use the [simplified comparative scale](https://docs.google.com/document/d/1jXayebAWnnNzsT3l6o72YPw4-YtiQaHQNuAi64eiguc/edit#heading=h.t9g7ilxdyv2) rather than the full comparative scales
    - When adjacent corners/sides are different than compared corner/side but similar in comparison to each other
        - “smaller/wider than adjacent corners”
        - “shorter/longer than adjacent sides”
    - When one adjacent corner or side is equal, call it out and do not describe the other adjacent corner/side
        - {{equal to one}} adjacent corner
        - {{equal to one}} adjacent side
    - When adjacent corners/sides are all together different from the compared corner/side and from each other describe the corner/side as “not equal to” adjacent corners/sides
        - {{not equal to} adjacent corners
        - {{not equal to}} adjacent sides
        - Note: See table for simplified comparative scales for describing [adjacent corners](https://docs.google.com/document/d/1jXayebAWnnNzsT3l6o72YPw4-YtiQaHQNuAi64eiguc/edit#heading=h.t9g7ilxdyv2) and [adjacent sides](https://docs.google.com/document/d/1jXayebAWnnNzsT3l6o72YPw4-YtiQaHQNuAi64eiguc/edit#heading=h.e8pilbnrr3pb)

**Comment by jessegreenberg** [comment]
The above commit adds object responses for the vertices on `focus`. Ill work on sides next.

**Comment by jessegreenberg** [comment]
Its a little difficult to see what is being changed in https://github.com/phetsims/quadrilateral/issues/130#issuecomment-1111365809 but I think it is 
>When “equal to opposite” corner/side and “equal to adjacent” corner/side - CAN we leave off the third “equal”
> I think we might need to call out when sides start in parallel. " {{parallel to and}} {{somewhat longer than}} opposite side"

**Comment by jessegreenberg** [comment]
One slight discrepancy in the google doc for adjacent side descriptions is that sometimes I see "adjacent equal sides" and other times I see "equal adjacent sides".

Over slack @terracoda and I discussed and decided it should be
>it should be "equal adjacent sides" and "equal adjacent corners".

We also decided to add the `{{parallel to and}}` phrase for parallel sides to all shapes (not just non-parallelograms) for now. These can be changed later any time.

**Comment by jessegreenberg** [comment]
I am not sure yet when object responses should be spoken but for review and debugging purposes we are speaking them on `focus` events and `down` events. This can be changed any time.

**Comment by jessegreenberg** [comment]
OK, Object responss for corners and sides are complete: https://phet-dev.colorado.edu/html/quadrilateral/1.0.0-dev.29/phet/quadrilateral_en_phet.html?showModelValues&printVoicingResponses&voicingInitiallyEnabled

@terracoda  back to you to review, let me know if there are any issues with this version!

**Comment by jessegreenberg** [comment]
@terracoda do you know how you created that shape? I tried to reproduce but I am not seeing the same thing trying to place the vertices:

![image](https://user-images.githubusercontent.com/6396244/165792135-7036e0a8-ef56-4e7e-8af1-ce9f9087449d.png)

I am guessing we need to fix this by having the detected shape name override the length comparison. Like "if it is a square/rhombus always described the opposite side as equal". But I am not sure how it got out of sync here in the first place, that would probably be good to understand.

**Comment by terracoda** [comment]
@jessegreenberg, I did mix up using both small and default step sizes, so that might be the issues. The small step size seems to allow 2 steps within tolerance.

**Comment by jessegreenberg** [comment]
I was not able to fix this but I think I see how this could happen with our angle tolerance interval and shape name detection function. @terracoda I believe this should be fixed.

**Comment by terracoda** [comment]
@jessegreenberg, Ok, I think you got the first strings implemented correctly. The corners sound perfect. No changes there.

 So sorry, I missed the parallel-ness on the adjacent sides phrases. You have parallel-ness working pretty good for the opposite side phrases.

I've updated all the examples, and the rules, and I think we only have 2 critical adjustments to make to the adjacent side descriptions, and 3 additional wording tweaks for other descriptions if they are possible.

**Adjacent sides** 
I missed the parallel parameter on both parallelograms and non-parallelograms.

1. For parallelograms the adjacent sides are always “parallel and equal” to each other _(not so for non-parallelograms)_, we need to add in the parallel parameter before the equal parameter, both just before "adjacent sides"
                - {{somewhat longer than}} {{parallel}} {{and}} {{equal}} adjacent sides

2. For non-parallelograms, we need to add the parallel parameter - whenever it is  true - before "adjacent sides". 
                -  longer than {{parallel}} adjacent sides _(for trapezoids)_
                - {{not equal to}} {{parallel}} adjacent sides _(for isosceles trapezoids)_

**Note:** the "equal" parameter is already working great. I can't think of a case for a non-parallelogram where the adjacent sides would be both "parallel and equal", so I think for non-parallelograms, it is an either-or situation. For example, _isosceles trapezoids_ already look like this:
        - {{a little longer than}} {{equal}} adjacent sides

**Opposite sides (tweak if possible)** 

3. For parallelograms a side is always “parallel and equal to” its opposite  _(not so for non-parallelograms)_. Is it possible to remove the first "to" and change "{{parallel to}} {{and}} {{equal to}} opposite side" to:
        - {{parallel}} {{and}} {{equal to}} opposite side 

**Note:** We need the "to" for non-parallelograms, and this description for non-parallelograms is currently working perfectly:
                -  {{parallel to}} {{and}} {{a little longer than}} opposite side

**When all things are equal (tweak if possible)**
When all things are equal, the word equal shows up three times. If possible can we leave off the third "equal"? That would mean 

4. For squares and rectangles changing:
        - {{equal to}} adjacent {{equal}} corners, **TO**
        - {{equal to}} adjacent corners
5. For squares and rhombuses changing:
        - {{equal to}} {{parallel}} and {{equal}} adjacent sides **TO**
        - {{equal to}} {{parallel}} adjacent sides

**Comment by terracoda** [comment]
@jessegreenberg, I need fix the previous comment, it is not quite right.

**Comment by jessegreenberg** [comment]
OK I think I have done everything in https://github.com/phetsims/quadrilateral/issues/130#issuecomment-1116514206 except for request 4. I tried for ~10 minutes but it was not going to be an easy tweak so I let it go. But I can return to it if it is important.

The updated examples in the google doc made these changes much easier for me to understand and implement, thank you for doing that @terracoda I appreciate it. Back to you to check this change.

**Comment by terracoda** [comment]
@jessegreenberg, the adjusted object responses sound great. I went through all the shapes.
